lightning/src/util/sweep.rs

```diff
@@ -463,39 +449,91 @@ where
 		self.sweeper_state.lock().unwrap().best_block
 	}
 
-	fn regenerate_spend_if_necessary(
-		&self, sweeper_state: &mut SweeperState,
-	) -> Option<Transaction> {
-		let cur_height = sweeper_state.best_block.height;
-		let cur_hash = sweeper_state.best_block.block_hash;
-		let filter_fn = |o: &TrackedSpendableOutput| {
-			if o.status.is_confirmed() {
-				// Don't rebroadcast confirmed txs.
-				return false;
-			}
+	/// Regenerates and broadcasts the spending transaction for any outputs that are pending
+	pub async fn regenerate_and_broadcast_spend_if_necessary_locked(&self) -> Result<(), ()> {
+		// Collect spendable output descriptors.
+		let respend_descriptors_clones: Vec<SpendableOutputDescriptor>;
+		let respend_descriptors: Vec<&SpendableOutputDescriptor>;
+		{
+			let mut sweeper_state = self.sweeper_state.lock().unwrap();
 
-			if o.status.is_delayed(cur_height) {
-				// Don't generate and broadcast if still delayed
-				return false;
+			// Prevent concurrent sweeping.
+			if sweeper_state.sweep_pending {
+				return Ok(());
 			}
 
-			if o.status.latest_broadcast_height() >= Some(cur_height) {
-				// Only broadcast once per block height.
-				return false;
+			let cur_height = sweeper_state.best_block.height;
+			let filter_fn = |o: &TrackedSpendableOutput| {
+				if o.status.is_confirmed() {
+					// Don't rebroadcast confirmed txs.
+					return false;
+				}
+
+				if o.status.is_delayed(cur_height) {
+					// Don't generate and broadcast if still delayed
+					return false;
+				}
+
+				if o.status.latest_broadcast_height() >= Some(cur_height) {
+					// Only broadcast once per block height.
+					return false;
+				}
+
+				true
+			};
+
+			// Clone first, otherwise we can't take references that outlive the lock.
+			respend_descriptors_clones =
+				sweeper_state.outputs.iter().filter(|o| filter_fn(*o)).map(|o| o.descriptor.clone()).collect();
+
+			respend_descriptors = respend_descriptors_clones.iter().collect();
+
+			if respend_descriptors.is_empty() {
+				// Nothing to do.
+				return Ok(());
 			}
 
-			true
-		};
+			// There is something to sweep. Block concurrent sweeps.
+			sweeper_state.sweep_pending = true;
+		}
+
+		// Request a new change address outside of the mutex to avoid the mutex crossing await.
+		let change_destination_script_result = self.change_destination_source.get_change_destination_script().await;
+
+		// Sweep the outputs.
+		{
+			let mut sweeper_state = self.sweeper_state.lock().unwrap();
 
-		let respend_descriptors: Vec<&SpendableOutputDescriptor> =
-			sweeper_state.outputs.iter().filter(|o| filter_fn(*o)).map(|o| &o.descriptor).collect();
+			// Always allow a new sweep after this spend.
+			sweeper_state.sweep_pending = false;
 
-		if respend_descriptors.is_empty() {
-			// Nothing to do.
-			return None;
+			let change_destination_script = change_destination_script_result?;
```

Comment:
Todo: address risk of getting a tx with a new address every block.

Reply:
Traced through BDK a bit. It seems that there is only inflation if we actually use the address in a tx. It won't blindly regenerate addresses when called. But will double check this.